generate_data/jzarr/src/main/java/zarr_implementations/jzarr/App.java

```diff
@@ -0,0 +1,111 @@
+package zarr_implementations.jzarr;
+
+import com.bc.zarr.ArrayParams;
+import com.bc.zarr.CompressorFactory;
+import com.bc.zarr.DataType;
+import com.bc.zarr.ZarrArray;
+import com.bc.zarr.ZarrGroup;
+
+import javax.imageio.ImageIO;
+import java.awt.*;
+import java.awt.image.BufferedImage;
+import java.awt.image.DataBuffer;
+import java.awt.image.DataBufferByte;
+import java.awt.image.DataBufferInt;
+import java.awt.image.WritableRaster;
+import java.io.ByteArrayOutputStream;
+import java.io.File;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.Arrays;
+import java.util.stream.IntStream;
+
+
+public class App {
+
+    enum Compression {
+        raw("null"),
+        zlib("zlib"),
+        blosc("blosc");
+
+        private final String value;
+
+        private Compression(final String value) {
+            this.value = value;
+        }
+
+        @Override
+        public String toString() {
+            return value;
+        }
+    }
+
+    // NOTE for now we use 100, 100, 1 as block-size in all examples
+    // maybe it's a better idea to make this more irregular though
+    private static final int WIDTH = 512;
+    private static final int HEIGHT = 512;
+    private static final int CHANNELS = 3;
+    private static final int[] CHUNKS = new int[]{100, 100, 1};
+    private static final int[] SHAPE = new int[] {WIDTH, HEIGHT, CHANNELS};
+    private static final Path IN_PATH = Paths.get("..", "..", "data", "reference_image.png");
+    private static final Path OUT_PATH = Paths.get("..", "..", "data", "jzarr_flat.zr");
+
+    private static int[] getData() throws IOException {
+        final BufferedImage image = ImageIO.read(new File(IN_PATH.toString()));
+        int[] result = new int[WIDTH * HEIGHT * CHANNELS];
+        for (int i = 0; i < WIDTH; i++) {
+            for (int j = 0; j < HEIGHT; j++) {
+                Color color = new Color(image.getRGB(i, j));
+                int index = (WIDTH*3*j) + (3*i);
+                result[index + 0] = color.getRed();
+                result[index + 1] = color.getGreen();
+                result[index + 2] = color.getBlue();
+            }
+        }
+        return result;
+    }
+
+    public static void main(String args[]) throws Exception {
+
+        if (args.length != 0 && args.length != 3) {
+            System.out.println("usage: App");
+            System.out.println("usage: App -verify fpath dsname");
+            System.exit(2);  // EARLY EXIT
+        } else if (args.length == 3) {
+            String fpath = args[1];
+            String dsname = args[2];
+            ZarrArray verification = ZarrGroup.open(fpath).openArray(dsname);
+            int[] shape = verification.getShape();
+            if (!Arrays.equals(new int[]{}, shape)) {
```

Comment:
Maybe I am reading this wrong, but why does it look like this line is comparing a new empty array to shape? I was expecting `Array.equals(SHAPE, shape)`? (I am not a Java dev)

Am I correct that when called with arguments, this program only checks the shape, but not the actual image contents?

What do you think about writing out a NumPy `.npy` or `.npz` file from Java? (A quick search indicates there [is at least one library available for this](https://stackoverflow.com/questions/18727600/writing-npy-numpy-binary-format-from-java) ([GitHub](https://github.com/JetBrains-Research/npy)). That is what the xtensor-zarr executable does (using an .npz writer that was already available in `xtensor-io`). The array comparison to the reference can then done in Python. Or maybe it is just easier to do the verification in java itself instead?

Reply:
> Maybe I am reading this wrong, but why does it look like this line is comparing a new empty array to shape? I was expecting `Array.equals(SHAPE, shape)`? (I am not a Java dev)

_Argh_. No, A) you're completely right. I added this before the troubles with `<u1` began in order to force a failure so B) why isn't it forcing a failure?!



> What do you think about writing out a NumPy `.npy` or `.npz` file from Java?

I'm personally skeptical. I think we're going to want to increase the amount of verification done in each language and it seems like that will hit limits via NumPy. But I assume I'm going to lose this one to you and @constantinpape 🙂

Reply:
see: 7883082fdc38313d61a1b4e6940bace3bdcf158c

Reply:
> I'm personally skeptical. I think we're going to want to increase the amount of verification done in each language and it seems like that will hit limits via NumPy. But I assume I'm going to lose this one to you and @constantinpape slightly_smiling_face

I am pretty agnostic on this one. I think it's fine if you validate in java. But we should somehow return the correct error codes to the python test to populate the test summary. (I haven't checked the rest of this PR yet, so maybe you are doing that already ;)).

Reply:
> why isn't it forcing a failure?!

I opened https://github.com/joshmoore/zarr_implementations/pull/1 with a simplified way of calling the subprocess that I am sure is actually running it (it takes a few seconds per test case). That one will fail if the exit code is not zero.

Reply:
I also do not have a problem with just doing the validation in Java